pilot/pkg/serviceregistry/kube/controller/ambient/statusqueue/queue_test.go

```diff
@@ -126,3 +127,99 @@ func TestQueue(t *testing.T) {
 	}})
 	expectConditions("conds", map[string]bool{"t2": true})
 }
+
+func TestQueueLeaderElection(t *testing.T) {
+	notifier := model.NewActiveNotifier(false)
+	q := statusqueue.NewQueue(notifier)
+	c := kube.NewFakeClient()
+	svc := kclient.New[*v1.Service](c)
+	svcs := krt.WrapClient[*v1.Service](svc)
+	col := krt.NewCollection(svcs, func(ctx krt.HandlerContext, i *v1.Service) *serviceStatus {
+		conds := model.ConditionSet{
+			model.ConditionType("t1"): nil,
+		}
+		for _, set := range strings.Split(i.Annotations["conditions"], ",") {
+			k, v, _ := strings.Cut(set, "=")
+			conds[model.ConditionType(k)] = &model.Condition{Status: v == "true", Reason: "some reason"}
+		}
+		return &serviceStatus{
+			Target: model.TypedObject{
+				NamespacedName: config.NamespacedName(i),
+				Kind:           kind.Service,
+			},
+			Conditions: conds,
+		}
+	})
+	statusqueue.Register(q, "services", col, func(status serviceStatus) (kclient.Patcher, []string) {
+		return kclient.ToPatcher(svc), nil
+	})
+	clienttest.Wrap(t, svc).Create(&v1.Service{ObjectMeta: metav1.ObjectMeta{Name: "none", Namespace: "default"}})
+	clienttest.Wrap(t, svc).Create(&v1.Service{ObjectMeta: metav1.ObjectMeta{
+		Name:        "conds",
+		Namespace:   "default",
+		Annotations: map[string]string{"conditions": "t1=true"},
+	}})
+	stop := test.NewStop(t)
+	c.RunAndWait(stop)
+	go q.Run(stop)
+
+	expectConditions := func(name string, conds map[string]bool) {
+		t.Helper()
+		retry.UntilSuccessOrFail(t, func() error {
+			s := svc.Get(name, "default")
+			if s == nil {
+				return fmt.Errorf("object not found")
+			}
+			allHave := slices.Map(s.Status.Conditions, func(t metav1.Condition) string {
+				return t.Type
+			})
+			have := slices.GroupUnique(s.Status.Conditions, func(t metav1.Condition) string {
+				return t.Type
+			})
+
+			for wc, wv := range conds {
+				realCond, f := have[wc]
+				if !f {
+					return fmt.Errorf("expected condition %q, had %v", wc, allHave)
+				}
+				delete(have, wc)
+				if (realCond.Status == metav1.ConditionTrue) != wv {
+					return fmt.Errorf("expected condition %q to be %v, got %v", wc, realCond.Status, wv)
+				}
+			}
+			if len(have) > 0 {
+				return fmt.Errorf("unexpected conditions, wanted %v, got %v", maps.Keys(conds), allHave)
+			}
+			return nil
+		}, retry.Timeout(time.Second*2))
+	}
+
+	// Initial state: no conditions
+	expectConditions("none", nil)
+	expectConditions("conds", nil)
+
+	// Win the leader election, should run existing state
+	notifier.StoreAndNotify(true)
+	expectConditions("conds", map[string]bool{"t1": true})
+
+	// Signal we lost the leader election
+	notifier.StoreAndNotify(false)
+	// Update should be ignored...
+	clienttest.Wrap(t, svc).CreateOrUpdate(&v1.Service{ObjectMeta: metav1.ObjectMeta{
+		Name:        "conds",
+		Namespace:   "default",
+		Annotations: map[string]string{"conditions": "t2=true"},
+	}})
```

Comment:
oops, forgot this one. we should check that updates don't write status when we are not active
```suggestion
	}})
	expectConditions("conds", map[string]bool{"t1": true})
```

Reply:
I didn't do it for a reason (which I meant to document, oops), in that it was racy -- fixed in the latest commit though, found a way around it. (the bug was in the test, not code)